main.rb

```
@@ -0,0 +1,198 @@
+$plan_id = env_has_key('AC_TESTINIUM_PLAN_ID')
+$ac_max_failure_percentage = (get_env_variable('AC_TESTINIUM_MAX_FAIL_PERCENTAGE') || 0).to_i
+$company_id = get_env_variable('AC_TESTINIUM_COMPANY_ID')
+$env_file_path = env_has_key('AC_ENV_FILE_PATH')
+$each_api_max_retry_count = env_has_key('AC_TESTINIUM_MAX_API_RETRY_COUNT').to_i
+timeout = env_has_key('AC_TESTINIUM_TIMEOUT').to_i
+date_now = DateTime.now
+$end_time = date_now + Rational(timeout, MINUTES_IN_A_DAY)
+$time_period = 30
+
+def get_parsed_response(response)
+  JSON.parse(response, symbolize_names: true)
+rescue JSON::ParserError, TypeError => e
+  puts "\nJSON expected but received: #{response}".red
+  puts "Error Message: #{e}".red
+  exit(1)
+end
+
+def calc_percent(numerator, denominator)
+  if !(denominator >= 0)
+    puts "Invalid numerator or denominator numbers".red
+    exit(1)
+  elsif denominator == 0
+    return 0
+  else
+    return numerator.to_f / denominator.to_f * 100.0
+  end
+end
+
+def check_timeout()
+  puts "Checking timeout...".yellow
+  now = DateTime.now
+
+  if now > $end_time
+    puts "Timeout exceeded! If you want to allow more time, please increase the AC_TESTINIUM_TIMEOUT input value.".red
+    exit(1)
+  end
+end
+
+def is_count_less_than_max_api_retry(count)
+  return count < $each_api_max_retry_count
+end
+
+def login()
+  puts "Logging in to Testinium...".yellow
+  uri = URI.parse('https://account.testinium.com/uaa/oauth/token')
+  token = 'dGVzdGluaXVtU3VpdGVUcnVzdGVkQ2xpZW50OnRlc3Rpbml1bVN1aXRlU2VjcmV0S2V5'
+  count = 1
+
+  while is_count_less_than_max_api_retry(count)
+    check_timeout()
+    puts "Signing in. Attempt: #{count}".blue
+
+    req = Net::HTTP::Post.new(uri.request_uri, { 'Content-Type' => 'application/json', 'Authorization' => "Basic #{token}" })
+    req.set_form_data({ 'grant_type' => 'password', 'username' => $username, 'password' => $password })
+    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
```

As far as I can see, only success and unauthorized are handled when making an HTTP request, but the API can throw another unexpected exceptional error during an HTTP request. Other errors are not handled. Can a begin/rescue structure be used to handle HTTP request errors better? What do you think? @ftmzhrsckldr

---

And also I see the same HTTP request in 4 different function. We can extract a new common function I think. For example, what do you think?

```
def send_http_request(req, uri, attempt)
 begin 
  Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
  end
 rescue standartError => e
puts "HTTP request failed on attempt #{atempt}: #{e.message}"
end
end
```

---

@boz These parts will change when I integrate Enterprise APIs (in 1 week) so I recommend not making changes here for now.